R/obic_score.R

```diff
@@ -59,4 +93,62 @@ obic_score <- function(dt.ind) {
 #' This table defines the weighting factors (ranging between 0 and 1) of indicator values to calculate integrated scores. 
 #' This table is used internally in \code{\link{obic_score}}
 #' 
-"weight.obic"
\ No newline at end of file
+"weight.obic"
+
+#' Score the scoring for the OBI on a relative scale
+#' 
+#' This function calculates the relative scoring based on the score calculated by \code{\link{score_absolute}}
+#' 
+#' @param dt.score.abs (data.table) The table containing the score calculated by \code{\link{score_absolute}}
+#' 
+#' @import data.table
+#' 
+#' @export
+score_relative <- function(dt.score.abs) {
+  
+  # Check inputs
+  checkmate::assert_data_table(dt.score.abs)
+  
+  # make local copy
+  dt.score.abs <- copy(dt.score.abs)
+  
+  group_id = V1 = col.sel = waterstress = yield_depression = NULL
+  S_C_R = S_P_R = S_B_R = S_M_R = S_T_R = NULL
+  S_C_A = S_P_A = S_B_A = S_M_A = S_T_A = NULL
+  
+  # Join crop categories
+  crops.obic <- as.data.table(OBIC::crops.obic)
+  col.sel <- c("crop_code", "crop_waterstress")
+  dt.score <- merge(dt.score.abs, crops.obic[, mget(col.sel)], by.x = "B_LU_BRP", by.y = "crop_code", all.x = TRUE)
+
+  # Join the waterstress levels
+  waterstress.obic <- as.data.table(OBIC::waterstress.obic)
+  waterstress.obic[waterstress < 11, yield_depression := 1]
+  waterstress.obic[waterstress >= 11 & waterstress < 30, yield_depression := 2]
+  waterstress.obic[waterstress >= 30, yield_depression := 3]
+  dt.score <- merge(dt.score, waterstress.obic, by.x = c("crop_waterstress", "B_HELP_WENR", "B_GT"), by.y = c("cropname", "soilunit", "gt"))
+
+  # Select columns to base the ranking on
+  grouping <- c("YEAR", "yield_depression", "crop_waterstress", "B_BT_AK")
+  groups <- dt.score[, (count = .N), by = grouping]
+  groups[, group_id := 1:.N]
+  groups[V1 < 10, group_id := -1]
+
+  # Join the group_id to
+  col.sel <- c(grouping, "group_id")
+  dt.score <- merge(dt.score, groups[, mget(col.sel)], by = grouping, all.x = TRUE)
+
+  # Rank the absolute values and scale them between 1 and 0
+  dt.score[, S_C_R := frank(S_C_A), by = group_id]
+  dt.score[, S_C_R := (S_C_R - min(S_C_R, na.rm = TRUE)) / (max(S_C_R, na.rm = TRUE) - min(S_C_R, na.rm = TRUE)), by = group_id]
```

Comment:
is it possible that SCR has NA values?

Reply:
During testing there were some cases that without na.rm an error occured due to a NA in S_*_R. However, it could be obsolete now